### PR TITLE
Simplify reference model handling in DPO

### PR DESCRIPTION
**Pull request by albertvillanova**
Simplify reference model handling.

This PR refactors how the reference model (`ref_model`) is handled, making its usage more consistent and robust, especially in scenarios involving PEFT adapters and precomputing reference log probabilities. The main changes ensure that `self.ref_model` is always set appropriately and that logic for PEFT and non-PEFT models is unified and clearer, by removing the scattered `None`-checks and nested PEFT conditions.

### Solution

Now `self.ref_model` is always a callable model. For PEFT with no explicit ref model it is `self.model` itself; for non-PEFT with `precompute_ref_log_probs=True` and no explicit ref model it is also `self.model` (valid as reference since no training step has run yet). An explicit `ref_model` always takes precedence.

### Changes

**Reference model initialization and usage:**

* Refactored the initialization logic so that `self.ref_model` is always set to either the explicitly provided `ref_model` or to `self.model`, depending on the training setup (PEFT, precompute_ref_log_probs, or standard). This ensures consistent access throughout the class. 
* Updated all downstream usage to reference `self.ref_model` directly, removing conditional checks and simplifying logic for hashing, caching, and inference.

**PEFT adapter handling:**

* Unified and clarified the logic for using PEFT adapters as the reference model, including correct adapter switching and model unwrapping, both in loss computation and when computing reference log probabilities.

**Model preparation and callbacks:**

* Standardized model preparation (e.g., dropout disabling, accelerator preparation) to always operate on the correct reference model, and ensured callbacks receive the appropriate reference model instance. 

---

> [!NOTE]
> **Medium Risk**
> Changes core DPO reference forward and precompute paths (PEFT adapter switching, DeepSpeed precompute, post-training evaluate); behavior should be equivalent but mistakes would skew preference loss.
> 
> **Overview**
> **DPO reference policy handling is refactored** so `self.ref_model` is always a callable model (`explicit ref_model` or `ref_model or self.model`), replacing scattered `None` checks.
> 
> **Initialization** now rejects passing a `PeftModel` as `ref_model` (PEFT should be the train `model` only). A separate reference checkpoint is still loaded only for full fine-tuning when neither PEFT nor `precompute_ref_log_probs` applies; accelerator prep and `SyncRefModelCallback` run on that local `ref_model` before assignment.
> 
> **Downstream paths** (`_precompute_ref_logps`, `compute_ref_log_probs`, `_compute_loss`, Liger loss, `evaluate`) use `is_peft_model(self.ref_model)` and `self.ref_model is self.model` instead of `self.ref_model is None`, including PEFT reference forwards via `unwrap_model(self.ref_model)` + `use_adapter`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ab0e2f644f193a49b2fafa056c850e99f40e732f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1202,13 +1202,13 @@ def _compute_loss(self, model, inputs, return_outputs):
...
+                if is_peft_model(self.ref_model):
                     # When training a PEFT adapter, how we obtain the reference depends on the setup:
                     # - New adapter: disabling adapters yields the base model.
                     # - Re-training an existing adapter: an initial copy is loaded under the name "ref".
-                    model = self.accelerator.unwrap_model(model)
-                    with use_adapter(model, adapter_name="ref" if "ref" in model.peft_config else None):
-                        ref_outputs = self.model(**model_kwargs)
+                    ref_model = self.accelerator.unwrap_model(self.ref_model)
+                    with use_adapter(ref_model, adapter_name="ref" if "ref" in ref_model.peft_config else None):
+                        ref_outputs = self.ref_model(**model_kwargs)
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py]: I think passing an explicit PEFT `ref_model` is an exotic edge case, probably never intended, and wasn't explicitly supported before either.

What do you think, @qgallouedec?

[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py]: I think we should raise an error in this case: 8ad8e09d

[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py]: Alternatively, if we want to support this edge case, I could modify the condition:
```diff
- if is_peft_model(self.ref_model):
+ if is_peft_model(self.ref_model) and self.ref_model is self.model:
```

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -549,6 +549,11 @@ def __init__(
                 "`model` and `ref_model` cannot be the same object. In most cases you should omit `ref_model` and "
                 "we'll initialize it to a copy of `model` for you."
             )
+        if ref_model is not None and is_peft_model(ref_model):
+            raise ValueError(
+                "`ref_model` cannot be a `PeftModel`. If you want to use a `PeftModel` as the reference model, please "
+                "pass it as `model` and set `ref_model=None`."
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py]:
```suggestion
                "pass it as `model` and don't pass anything for `ref_model`."
```

**Review by qgallouedec — APPROVED** [review]
thanks! agree passing a peft model as ref is exotic, we don't have to support this case.

**Comment by albertvillanova** [comment]
After rethinking about this PR, I am not totally sure about this approach: some cases are simpler but others are not so clear!

**Comment by qgallouedec** [comment]
Right, let's drop it.
Looking again, the refactor trades one sentinel for an attribute that means three things, and each call site needs a different test to tell them apart; tt's behaviour-preserving but not simpler.

The piece that would be worth keeping is the guard rejecting a `PeftModel` as `ref_model`. 

To keep the queue tidy, I'll close #5877 with this one.
